src/js/tls.js

```diff
@@ -15,184 +15,224 @@
 
-  if (opts.ALPNProtocols && !util.isBuffer(opts.ALPNProtocols)) {
-    throw new TypeError('TLSSocket - options.ALPNProtocols must be a buffer.');
-  }
+TLSSocket.prototype.destroy = function() {
+  this._socket.destroy();
+};
 
-  if (opts.SNICallback && !util.isFunction(opts.SNICallback)) {
-    throw new TypeError('TLSSocket - options.SNICallback must be a function.');
-  }
+TLSSocket.prototype.destroySoon = function() {
+  this._socket.destroySoon();
+};
 
-  if (opts.session && !util.isBuffer(opts.session)) {
-    throw new TypeError('TLSSocket - options.session should be a buffer.');
+TLSSocket.prototype.onconnect = function() {
+  var self = this._tlsSocket;
+  if (self._ended) {
+    return;
   }
+  self._read(null);
 };
 
-Tls.TLSSocket.prototype.address = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.encrypted = function() {
+  return true;
 };
 
-Tls.TLSSocket.prototype.disableRenegotiation = function() {
-  this.disableRenegotiation = true;
+TLSSocket.prototype.address = function() {
+  return this._socket.address();
 };
 
-Tls.TLSSocket.prototype.encrypted = function() {
-  return this.encrypted;
+TLSSocket.prototype.localAddress = function() {
+  return this._socket.address().address;
 };
 
-Tls.TLSSocket.prototype.getCipher = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.ondata = function(data) {
+  var self = this._tlsSocket;
+  if (self._ended) {
+    return;
+  }
+  self._read(data);
 };
 
-Tls.TLSSocket.prototype.getEphemeralKeyInfo = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.onerror = function(error) {
+  this._tlsSocket.emit('error', error);
 };
 
-Tls.TLSSocket.prototype.getPeerCertificate = function(/* detailed */) {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.onclose = function() {
+  this._tlsSocket.emit('close');
 };
 
-Tls.TLSSocket.prototype.getProtocol = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.onfinish = function() {
+  this._tlsSocket.emit('finish');
 };
 
-Tls.TLSSocket.prototype.getSession = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.onend = function() {
+  this._tlsSocket.emit('end');
 };
 
-Tls.TLSSocket.prototype.getTLSTicket = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.onwrite = function(data) {
+  return this._socket.write(data);
 };
 
-Tls.TLSSocket.prototype.write = function(message) {
-  return native.write(message);
+TLSSocket.prototype.onread = function(chunk) {
+  this.emit('data', chunk);
 };
 
-Tls.TLSSocket.prototype.renegotiate = function(options, callback) {
-  if (!util.isObject(options)) {
-    throw new TypeError('options should be an object');
-  }
+TLSSocket.prototype.onhandshakedone = function(error, authorized) {
+  this.authorized = authorized;
 
-  if (callback && !util.isFunction(callback)) {
-    throw new TypeError('callback should be a function');
+  var server = this._tlsServer;
```

Comment:
Please use `this._server` instead of `this._tlsServer`. This will be needed for HTTPS. Also remove the deletion of it, because it is used later.